numba/tests/test_fancy_indexing.py

```diff
@@ -3,183 +3,207 @@
-        """
-        Generate basic index tuples with 0 to *maxdim* items.
-        """
-        # Note integers can be considered advanced indices in certain
-        # cases, so we avoid them here.
-        # See "Combining advanced and basic indexing"
-        # in http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html
-        if many:
-            choices = [slice(None, None, None),
-                       slice(1, N - 1, None),
-                       slice(0, None, 2),
-                       slice(N - 1, None, -2),
-                       slice(-N + 1, -1, None),
-                       slice(-1, -N, -2),
-                       ]
-        else:
-            choices = [slice(0, N - 1, None),
-                       slice(-1, -N, -2)]
-        for ndim in range(maxdim + 1):
-            for tup in itertools.product(choices, repeat=ndim):
-                yield tup
-
-    def generate_advanced_index_tuples(self, N, maxdim, many=True):
-        """
-        Generate advanced index tuples by generating basic index tuples
-        and adding a single advanced index item.
-        """
-        # (Note Numba doesn't support advanced indices with more than
-        #  one advanced index array at the moment)
-        choices = list(self.generate_advanced_indices(N, many=many))
-        for i in range(maxdim + 1):
-            for tup in self.generate_basic_index_tuples(N, maxdim - 1, many):
-                for adv in choices:
-                    yield tup[:i] + (adv,) + tup[i:]
-
-    def generate_advanced_index_tuples_with_ellipsis(self, N, maxdim, many=True):
-        """
-        Same as generate_advanced_index_tuples(), but also insert an
-        ellipsis at various points.
-        """
-        for tup in self.generate_advanced_index_tuples(N, maxdim, many):
-            for i in range(len(tup) + 1):
-                yield tup[:i] + (Ellipsis,) + tup[i:]
-
-    def check_getitem_indices(self, arr, indices):
-        pyfunc = getitem_usecase
-        cfunc = jit(nopython=True)(pyfunc)
+    # Every case has exactly one, one-dimensional array,
+    # Otherwise it's not fancy indexing
+    shape = (5, 6, 7, 8, 9, 10)
+    indexing_cases = [
+        # Slices + Integers
+        (slice(4, 5), 3, np.array([0,1,3,4,2]), 1),
+        (3, np.array([0,1,3,4,2]), slice(None), slice(4)),
+
+        # Ellipsis + Integers
+        (Ellipsis, 1, np.array([0,1,3,4,2])),
+        (np.array([0,1,3,4,2]), 3, Ellipsis),
+
+        # Ellipsis + Slices + Integers
+        (Ellipsis, 1, np.array([0,1,3,4,2]), 3, slice(1,5)),
+        (np.array([0,1,3,4,2]), 3, Ellipsis, slice(1,5)),
+
+        # Boolean Arrays + Integers
+        (slice(4, 5), 3,
+         np.array([True, False, True, False, True, False, False]),
+         1),
+        (3, np.array([True, False, True, False, True, False]),
+         slice(None), slice(4)),
+    ]
+
+    rng = np.random.default_rng(1)
+
+    def generate_random_indices(self):
```

Comment:
A few points regarding why I gave a last try to making this function rather than completely switching to old testing framework, There are following issues with the old testing logic:

1. The old framework tests on an array having only one size across dimensions on top of only having 3 dimensions. (its shape is simply 4x4x4).
2. Nearly ~400 indices are being generated as all the possible combination of the slices (which is unnecessary in my opinion) and a single array index which is same across all the indices (which is limited).
3. The testing for Ellipsis object is actually not testing the ellipsis object :-(, we are adding the ellipsis as a fourth index in a three dimensional array, so it's basically being dropped rather than being actually expanded into slices.
4. Adding dimensions to array to fix the above would basically add a huge number of cases in old logic, a few thousand indexing cases which will take much longer to run and is even more unnecessary. 

This function pretty much covers most of the needed logic and cases used previously, reducing the cases to ~80 (20 of each type) with the new ones being added separately.